### PR TITLE
IFP: Multi height cell support

### PR DESCRIPTION
**Pull request by mina1460**
This PR would serve as the foundation for supporting multi-height cells.

It also includes support for removing the need for the -site name flag for init_floorplan command. It would still work with and without it.

### CODE REVIEW

File: src/ifp/include/ifp/InitFloorplan.hh
Comment (maliberty) [inline src/ifp/include/ifp/InitFloorplan.hh:40]: I don't see a usage of vector in this header.  It is needed elsewhere please add it to the file(s) that use it.

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:402]: A warning would be good here as this would be unexpected.

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:419]: Use integers here not doubles.  Check with the modulo operator not ceil/floor.

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:422]: Why is this a TODO?

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:423]: Give the site names too

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:427]: you'll never get here as error() throws an exception.  The erase wouldn't be safe while iterating anyways

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:433]: Please add a unit test that inserts double height rows.

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:413]: site_vec[0] is invalid if sites.empty()

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:188]: rm, not reached

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:191]: const

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:201]: The core size won't change with sites so this check should be hoisted.

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:449]: any even factor not just 2 (eg quad height cells)

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/ifp/src/InitFloorplan.cc
Comment (mina1460) [reply src/ifp/src/InitFloorplan.cc:422]: I didn't know how I should pick an error code, I thought I should document the code somewhere as each code seems to be unique for a message. Should I just pick the first available number?

Comment (maliberty) [reply src/ifp/src/InitFloorplan.cc:422]: Use the next available from ifp/messages.txt (ie 40)

File: src/ifp/src/InitFloorplan.cc
Comment (maliberty) [reply src/ifp/src/InitFloorplan.cc:433]: This still needs addressing

**Comment by QuantamHD** [comment]
@mina1460 Would you mind rebasing to master I introduced a clang-tidy action that leaves comments on your PR that I would like to see here https://github.com/The-OpenROAD-Project/OpenROAD/pull/2980

**Comment by mina1460** [comment]
@QuantamHD Done

**Comment by QuantamHD** [comment]
Thanks!

File: src/ifp/test/mock.tcl
Comment (maliberty) [inline src/ifp/test/mock.tcl:7]: The test needs to do something to validate the results.  Likely you want a write_def and a .defok file.  Please see other tests for examples.

Comment (mina1460) [reply src/ifp/test/mock.tcl:7]: yes, absolutely. I just wanted to verify that the lef/def files and the mock macros are OK and this is indeed testing a double height cell before spending time on a defok file.

Should I proceed with it?

Comment (maliberty) [reply src/ifp/test/mock.tcl:7]: yes

File: src/ifp/test/mock.def
Comment (maliberty) [inline src/ifp/test/mock.def]: Why are there rows in the input DEF?  The goal of ifp is to build the rows.  You could just read a .v for this test which would better mirror ORFS (ie the yosys output).

Comment (mina1460) [reply src/ifp/test/mock.def]: makes sense. I will fix it

**Comment by QuantamHD** [comment]
@maliberty @vvbandeira I think @mina1460 is getting a high number of notifications from the sync upstream, because of the waiting for approval status. We should probably add him to the list of approved workflow runners.

**Comment by maliberty** [comment]
@QuantamHD  once his first PR is merged he will automatically be added.  I've never had to do it manually.

**Comment by maliberty** [comment]
The CI failures need to be addressed before this can be merged

**Comment by mina1460** [comment]
Hi @maliberty 

All CI/CD failures are now resolved. 

I modified the updateVoltageDomain and removed the site from it so it is no longer needed. As you said, this had the potential of cutting double-height rows into single-height rows, but I don't think this is a problem with this modification anymore.

**Review by maliberty — APPROVED** [review]

**Pull request merged**
